# Import libraries

In [1]:
#import necessary libraries
#import pandas
import pandas as pd

#import numpy
import numpy as nm 


# Import dataset files

In [2]:
#import csv files
ad = pd.read_csv('actual_duration.csv')
ar = pd.read_csv('appointments_regional.csv')

#import excel file
nc = pd.read_excel('national_categories.xlsx')

# Sense check actual_duration.csv
check column names, total number of rows and colums, data types
missing number of values, first 5 records

In [3]:
#check number of rows and columns
print(ad.shape)

#check column names
print(ad.columns)

#check data type
ad.dtypes

(137793, 8)
Index(['sub_icb_location_code', 'sub_icb_location_ons_code',
       'sub_icb_location_name', 'icb_ons_code', 'region_ons_code',
       'appointment_date', 'actual_duration', 'count_of_appointments'],
      dtype='object')


sub_icb_location_code        object
sub_icb_location_ons_code    object
sub_icb_location_name        object
icb_ons_code                 object
region_ons_code              object
appointment_date             object
actual_duration              object
count_of_appointments         int64
dtype: object

In [4]:
#check right data type for actual_duration.csv
ad.head()

sub_icb_location_code sub_icb_location_ons_code  \
0                   00L                 E38000130   
1                   00L                 E38000130   
2                   00L                 E38000130   
3                   00L                 E38000130   
4                   00L                 E38000130   

                        sub_icb_location_name icb_ons_code region_ons_code  \
0  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
1  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
2  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
3  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
4  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   

  appointment_date         actual_duration  count_of_appointments  
0        01-Dec-21           31-60 Minutes                    364  
1        01-Dec-21           21-30 Minutes                    619  
2        01-Dec-21            6-10 Minutes                   1698  
3        01-Dec-21  Unknown / Data Quality                   1277  
4        01-Dec-21           16-20 Minutes                    730

In [5]:
#check for number of missing values for actual_duration.csv
ad.isna().sum().sum()

0

# Descriptive statistics and metadata
# actual_duration.csv

In [6]:
#determine descriptive statistics for entire Data frame
stats_ad = ad.describe(include='all')
print (stats_ad)

       sub_icb_location_code sub_icb_location_ons_code  \
count                 137793                    137793   
unique                   106                       106   
top                      26A                 E38000239   
freq                    1484                      1484   
mean                     NaN                       NaN   
std                      NaN                       NaN   
min                      NaN                       NaN   
25%                      NaN                       NaN   
50%                      NaN                       NaN   
75%                      NaN                       NaN   
max                      NaN                       NaN   

                    sub_icb_location_name icb_ons_code region_ons_code  \
count                              137793       137793          137793   
unique                                106           42               7   
top     NHS Norfolk and Waveney ICB - 26A    E54000057       E40000010   
freq   

In [7]:
ad.describe()

count_of_appointments
count          137793.000000
mean             1219.080011
std              1546.902956
min                 1.000000
25%               194.000000
50%               696.000000
75%              1621.000000
max             15400.000000

In [8]:
ad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137793 entries, 0 to 137792
Data columns (total 8 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   sub_icb_location_code      137793 non-null  object
 1   sub_icb_location_ons_code  137793 non-null  object
 2   sub_icb_location_name      137793 non-null  object
 3   icb_ons_code               137793 non-null  object
 4   region_ons_code            137793 non-null  object
 5   appointment_date           137793 non-null  object
 6   actual_duration            137793 non-null  object
 7   count_of_appointments      137793 non-null  int64 
dtypes: int64(1), object(7)
memory usage: 8.4+ MB


# Sense check appointments_regional.csv
check column names, total number of rows and colums, data types
missing number of values, first 5 records

In [9]:
#check number of rows and columns
print(ar.shape)

#check column names
print(ar.columns)

#check data type
ar.dtypes

(596821, 7)
Index(['icb_ons_code', 'appointment_month', 'appointment_status', 'hcp_type',
       'appointment_mode', 'time_between_book_and_appointment',
       'count_of_appointments'],
      dtype='object')


icb_ons_code                         object
appointment_month                    object
appointment_status                   object
hcp_type                             object
appointment_mode                     object
time_between_book_and_appointment    object
count_of_appointments                 int64
dtype: object

In [10]:
#check right data type for appointments_regional.csv
ar.head()

icb_ons_code appointment_month appointment_status hcp_type appointment_mode  \
0    E54000034           2020-01           Attended       GP     Face-to-Face   
1    E54000034           2020-01           Attended       GP     Face-to-Face   
2    E54000034           2020-01           Attended       GP     Face-to-Face   
3    E54000034           2020-01           Attended       GP     Face-to-Face   
4    E54000034           2020-01           Attended       GP     Face-to-Face   

  time_between_book_and_appointment  count_of_appointments  
0                             1 Day                   8107  
1                    15  to 21 Days                   6791  
2                       2 to 7 Days                  20686  
3                    22  to 28 Days                   4268  
4                     8  to 14 Days                  11971

In [11]:
#check for number of missing values for appointments_regional.csv
ar.isna().sum().sum()

0

# Descriptive statistics and metadata
# appointments_regional.csv

In [12]:
#determine descriptive statistics for entire Data frame
stats_ar = ar.describe(include='all')
print (stats_ar)

       icb_ons_code appointment_month appointment_status  \
count        596821            596821             596821   
unique           42                30                  3   
top       E54000057           2020-03           Attended   
freq          54527             21350             232137   
mean            NaN               NaN                NaN   
std             NaN               NaN                NaN   
min             NaN               NaN                NaN   
25%             NaN               NaN                NaN   
50%             NaN               NaN                NaN   
75%             NaN               NaN                NaN   
max             NaN               NaN                NaN   

                    hcp_type appointment_mode  \
count                 596821           596821   
unique                     3                5   
top     Other Practice staff     Face-to-Face   
freq                  241557           180410   
mean                     NaN      

In [13]:
ar.describe()

count_of_appointments
count          596821.000000
mean             1244.601857
std              5856.887042
min                 1.000000
25%                 7.000000
50%                47.000000
75%               308.000000
max            211265.000000

In [14]:
ar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596821 entries, 0 to 596820
Data columns (total 7 columns):
 #   Column                             Non-Null Count   Dtype 
---  ------                             --------------   ----- 
 0   icb_ons_code                       596821 non-null  object
 1   appointment_month                  596821 non-null  object
 2   appointment_status                 596821 non-null  object
 3   hcp_type                           596821 non-null  object
 4   appointment_mode                   596821 non-null  object
 5   time_between_book_and_appointment  596821 non-null  object
 6   count_of_appointments              596821 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 31.9+ MB


# Sense check national_categories.xlsx
check column names, total number of rows and colums, data types
missing number of values, first 5 records

In [15]:
#check number of rows and columns
print(nc.shape)

#check column names
print(nc.columns)

#check data type
nc.dtypes


(817394, 8)
Index(['appointment_date', 'icb_ons_code', 'sub_icb_location_name',
       'service_setting', 'context_type', 'national_category',
       'count_of_appointments', 'appointment_month'],
      dtype='object')


appointment_date         datetime64[ns]
icb_ons_code                     object
sub_icb_location_name            object
service_setting                  object
context_type                     object
national_category                object
count_of_appointments             int64
appointment_month                object
dtype: object

In [16]:
#check right data type for national_categories.xlsx
nc.head()

appointment_date icb_ons_code                       sub_icb_location_name  \
0       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
1       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
2       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
3       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
4       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   

        service_setting            context_type  \
0  Primary Care Network  Care Related Encounter   
1                 Other  Care Related Encounter   
2      General Practice  Care Related Encounter   
3      General Practice  Care Related Encounter   
4      General Practice  Care Related Encounter   

                        national_category  count_of_appointments  \
0  Patient contact during Care Home Round                      3   
1                         Planned Clinics                      7   
2                              Home Visit                     79   
3              General Consultation Acute                    725   
4            Structured Medication Review                      2   

  appointment_month  
0           2021-08  
1           2021-08  
2           2021-08  
3           2021-08  
4           2021-08

In [17]:
#check for number of missing values for national_categories.xlsx
nc.isnull().sum().sum()

0

# Descriptive statistics and metadata
# national_categories.xlsx

In [18]:
#determine descriptive statistics for entire Data frame
stats_nc = nc.describe(include='all')
print (stats_nc)

C:\Users\maami\AppData\Local\Temp\ipykernel_21704\3059951789.py:2: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  stats_nc = nc.describe(include='all')


           appointment_date icb_ons_code              sub_icb_location_name  \
count                817394       817394                             817394   
unique                  334           42                                106   
top     2022-06-28 00:00:00    E54000057  NHS North West London ICB - W2U3Z   
freq                   3536        59180                              13007   
first   2021-08-01 00:00:00          NaN                                NaN   
last    2022-06-30 00:00:00          NaN                                NaN   
mean                    NaN          NaN                                NaN   
std                     NaN          NaN                                NaN   
min                     NaN          NaN                                NaN   
25%                     NaN          NaN                                NaN   
50%                     NaN          NaN                                NaN   
75%                     NaN          NaN            

In [19]:
nc.describe()

count_of_appointments
count          817394.000000
mean              362.183684
std              1084.576600
min                 1.000000
25%                 7.000000
50%                25.000000
75%               128.000000
max             16590.000000

In [20]:
nc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB


# answering questions
### How many locations in the Data set? 
106
### What are the five locations with the highest number of records?
NHS North West London ICB - W2U3Z              
NHS Kent and Medway ICB - 91Q                  
NHS Devon ICB - 15N                            
NHS Hampshire and Isle Of Wight ICB - D9Y0V    
NHS North East London ICB - A3A8R
### How many service settings, context types, national categories, and appointment statuses are there?
service settings - 5
context types - 3
national categories - 18
appointment statuses - 3

In [21]:
#find unique locations in the data set
loc_nc = nc.sub_icb_location_name.value_counts()
#print locations in descending order
print(loc_nc)

NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
                                               ...  
NHS North East and North Cumbria ICB - 00N      4210
NHS Lancashire and South Cumbria ICB - 02G      4169
NHS Cheshire and Merseyside ICB - 01V           3496
NHS Cheshire and Merseyside ICB - 01T           3242
NHS Greater Manchester ICB - 00V                2170
Name: sub_icb_location_name, Length: 106, dtype: int64


In [38]:
# count total number of unique locations
count_loc= len(loc_nc)
#print results
print("Count of locations:", count_loc)

Count of locations: 106


In [41]:
#view first 5 locations with the highest number of records
loc_nc.head()

NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
Name: sub_icb_location_name, dtype: int64

In [58]:
#find types of different service settings
service_set_nc = nc.service_setting.value_counts()
#print service settings
print(service_set_nc)

#count total number of service settings
count_ser_set = nc.service_setting.unique().size
#print results
print("Service settings: "+ str(count_ser_set))

General Practice             359274
Primary Care Network         183790
Other                        138789
Extended Access Provision    108122
Unmapped                      27419
Name: service_setting, dtype: int64
Service settings: 5


In [59]:
# find different context types
context_type = nc.context_type.value_counts()
# print outcome
print(context_type)
# count total number of different context types
count_cont = nc.context_type.unique().size
#print outcome
print("Context types: "+ str(count_cont))

Care Related Encounter    700481
Inconsistent Mapping       89494
Unmapped                   27419
Name: context_type, dtype: int64
Context types: 3


In [57]:
# find different national categories
nat_cat_nc = nc.national_category.value_counts()
# print outcome
print(nat_cat_nc)

#count different national categories
count_nat_cat = nc.national_category.unique().size
#print outcome
print("national categories: "+ str(count_nat_cat))

Inconsistent Mapping                                                   89494
General Consultation Routine                                           89329
General Consultation Acute                                             84874
Planned Clinics                                                        76429
Clinical Triage                                                        74539
Planned Clinical Procedure                                             59631
Structured Medication Review                                           44467
Service provided by organisation external to the practice              43095
Home Visit                                                             41850
Unplanned Clinical Activity                                            40415
Patient contact during Care Home Round                                 28795
Unmapped                                                               27419
Care Home Visit                                                        26644

In [56]:
#find appointment statuses
app_stat_ar = ar.appointment_status.value_counts()
#print outcome
print(app_stat_ar)
# count types of appointment status
count_app = ar.appointment_status.unique().size
#print outcome
print("Appointment statuses: "+ str(count_app))

Attended    232137
Unknown     201324
DNA         163360
Name: appointment_status, dtype: int64
Appointment statuses: 3
